modules/table/src/main/java/org/apache/ignite/internal/table/distributed/storage/InternalTableImpl.java

```diff
@@ -310,4 +344,117 @@ private int partId(BinaryRow row) {
                     return list;
                 });
     }
+
+    /** Parition scan publisher. */
+    private class PartitionScanPublisher implements Publisher<BinaryRow> {
+        /** List of subscriptions. */
+        private final List<PartitionScanSubscription> subscriptions;
+
+        /** {@link Publisher<BinaryRow>} that relatively notifies about partition rows.  */
+        private final RaftGroupService raftGrpSvc;
+
+        /**
+         * The constructor.
+         *
+         * @param raftGrpSvc {@link Publisher<BinaryRow>} that relatively notifies about partition rows.
+         */
+        PartitionScanPublisher(RaftGroupService raftGrpSvc) {
+            this.subscriptions = Collections.synchronizedList(new ArrayList<>());
+            this.raftGrpSvc = raftGrpSvc;
+        }
+
+        /** {@inheritDoc} */
+        @Override public void subscribe(Subscriber<? super BinaryRow> subscriber) {
+            PartitionScanSubscription subscription = new PartitionScanSubscription(subscriber);
+
+            subscriptions.add(subscription);
```

Comment:
`one subscriber per publisher` is actually a good idea not only for first iteration, [flow documentation](https://docs.oracle.com/javase/9/docs/api/java/util/concurrent/Flow.html) claims that
`it is better to instead define multiple Subscribers, each with its own Subscription`

` Because Subscriber method invocations for a given Flow.Subscription are strictly ordered, there is no need for these methods to use locks or volatiles unless a Subscriber maintains multiple Subscriptions (in which case it is better to instead define multiple Subscribers, each with its own Subscription). `